### PR TITLE
add declaration shadows matcher

### PR DESCRIPTION
**Pull request by brianyu28**
Addressing Issue https://github.com/cs50/help50/issues/3, 

Adding a matcher to `clang.py` to address `declaration shadows a local variable` error. This one's a little bit more involved than most of the others for a few reasons:

- it checks to see whether the error message is the result of a for loop, in which case the helper message needs to be different (because the student probably just used `,` instead of `;` in their loop definition).
- if the error wasn't generated from a for loop, then we try to get the line number of the variable's previous declaration if possible, in order to provide a more useful helper.

Separated this into its own branch per your comment on workflow.

### CODE REVIEW

File: helpers/clang.py [hunk 1]
```diff
@@ -1,5 +1,37 @@
...
+    # declaration shadows a local variable
+    matches = re.search(r"^([^:]+):(\d+):\d+: (?:warning|error): declaration shadows a local variable", lines[0])
+    if matches:
+        after = [
+            "On line {} of `{}`, it seems that you're trying to create a new variable which has already been created.".format(matches.group(2), matches.group(1))
+        ]
+        
+        # check to see if declaration shadowing is due to for loop with commas instead of semicolons
+        if len(lines) >= 2:
+            for_loop = re.search(r"^\s*for[(\s]", lines[1])
```
Comment (dmalan) [inline helpers/clang.py]: Hmm, sure the `[]` character class has the right semantics? Is `...for\s*\(` a closer match for what you're looking for?

File: helpers/clang.py [hunk 2]
```diff
@@ -1,5 +1,37 @@
...
+            if for_loop:
+                after.append("If you meant to create a for loop, be sure that each part of the for loop is separated with a semicolon (;) rather than a comma (,).")
```
Comment (dmalan) [inline helpers/clang.py]: ``for``

File: helpers/clang.py [hunk 3]
```diff
@@ -1,5 +1,37 @@
...
+                return (lines[0:2], after)
+        
+        # see if we can get the line number of the previous declaration of the varaible
```
Comment (dmalan) [inline helpers/clang.py]: varaible

File: helpers/clang.py [hunk 4]
```diff
@@ -1,5 +1,37 @@
...
+            if prev:
+                prev_declaration_line = prev.group(2)
+                prev_declaration_file = prev.group(1)
+        
+        omit_suggestion = "If you meant to use the variable you've already declared previously"
+        if prev_declaration_line and prev_declaration_file:
+            omit_suggestion += " (on line {} of `{}`)".format(prev_declaration_line, prev_declaration_file)
+        omit_suggestion += ", try getting rid of the data type of the variable on line {} of `{}`. You only need to include the data type when you first declare a variable.".format(matches.group(2), matches.group(1))
+        after.append(omit_suggestion)
+        after.append("Otherwise, if you did mean to declare a new variable, try changing its name to a name which hasn't been used yet.")
```
Comment (dmalan) [inline helpers/clang.py]: s/which/that/

**Comment by brianyu28** [comment]
Made tweaks per your comments, should be addressed in the most recent commit.

**Pull request merged**